From f465fcbd3402dc3ec1882c489ebf388e3e2f80fd Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 22 Nov 2012 21:30:59 -0500
Subject: [PATCH] Misc EMC Changes

Only spawn 1 ocelot ever from eggs.
Load custom permissions before plugins
Unload plugins after everything else
Make Durability enchant scale 0-80% reduction on Armor damage (up to level 5)
Zombies break down doors on normal
Update inventory on block place/interact cancellation
Don't update server.properties
Remove spigot update check
---
 src/main/java/net/minecraft/server/EntityOcelot.java             | 2 +-
 src/main/java/net/minecraft/server/ItemStack.java                | 9 +++++++++
 src/main/java/net/minecraft/server/MinecraftServer.java          | 2 ++
 src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java  | 2 +-
 src/main/java/net/minecraft/server/PropertyManager.java          | 2 +-
 src/main/java/org/bukkit/craftbukkit/CraftServer.java            | 3 ++-
 .../java/org/bukkit/craftbukkit/event/CraftEventFactory.java     | 2 ++
 7 files changed, 18 insertions(+), 4 deletions(-)

diff --git a/src/main/java/net/minecraft/server/EntityOcelot.java b/src/main/java/net/minecraft/server/EntityOcelot.java
index ea5adbe..33f9026 100644
--- a/src/main/java/net/minecraft/server/EntityOcelot.java
+++ b/src/main/java/net/minecraft/server/EntityOcelot.java
@@ -6,7 +6,7 @@ public class EntityOcelot extends EntityTameableAnimal {
 
     private PathfinderGoalAvoidTarget<EntityHuman> bo;
     private PathfinderGoalTempt bp;
-    public boolean spawnBonus = true; // Spigot
+    public boolean spawnBonus = false; // Spigot // EMC
 
     public EntityOcelot(World world) {
         super(world);
diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index 1984e7c..03391eb 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -356,7 +356,16 @@ public final class ItemStack {
                 int j = EnchantmentManager.getEnchantmentLevel(Enchantment.DURABILITY.id, this);
                 int k = 0;
 
+                j = Math.min(5, j); // EMC
                 for (int l = 0; j > 0 && l < i; ++l) {
+                    // EMC start - make Unbreaking scale to level for armor. 5 = 80% chance of no loss.
+                    if (getItem() instanceof ItemArmor) {
+                        if (random.nextFloat() >= 1 - (.8 * (j / 5))) {
+                            k++;
+                        }
+                        continue;
+                    }
+                    // EMC end
                     if (EnchantmentDurability.a(this, j, random)) {
                         ++k;
                     }
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index b189c19..b8bb2d2 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -453,11 +453,13 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
         // CraftBukkit end
         if (!this.N) {
             MinecraftServer.LOGGER.info("Stopping server");
+
             // CraftBukkit start
             if (this.server != null) {
                 this.server.disablePlugins();
             }
             // CraftBukkit end
+
             if (this.ap() != null) {
                 this.ap().b();
             }
diff --git a/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java b/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
index fdc77cb..4948867 100644
--- a/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
+++ b/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
@@ -62,7 +62,7 @@ public class PathfinderGoalBreakDoor extends PathfinderGoalDoorInteract {
             this.h = i;
         }
 
-        if (this.g == 240 && this.a.world.getDifficulty() == EnumDifficulty.HARD) {
+        if (this.g == 240 && this.a.world.getDifficulty().ordinal() >= EnumDifficulty.NORMAL.ordinal()) { // EMC
             // CraftBukkit start
             if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreakDoorEvent(this.a, this.b.getX(), this.b.getY(), this.b.getZ()).isCancelled()) {
                 this.c();
diff --git a/src/main/java/net/minecraft/server/PropertyManager.java b/src/main/java/net/minecraft/server/PropertyManager.java
index 0549a0f..8b3e13c 100644
--- a/src/main/java/net/minecraft/server/PropertyManager.java
+++ b/src/main/java/net/minecraft/server/PropertyManager.java
@@ -72,7 +72,7 @@ public class PropertyManager {
 
         try {
             // CraftBukkit start - Don't attempt writing to file if it's read only
-            if (this.file.exists() && !this.file.canWrite()) {
+            if (this.file.exists()) { // EMC - Never overwrite the file
                 return;
             }
             // CraftBukkit end
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 6174417..3b634a8 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -307,6 +307,7 @@ public final class CraftServer implements Server {
         if (type == PluginLoadOrder.STARTUP) {
             helpMap.clear();
             helpMap.initializeGeneralTopics();
+            loadCustomPermissions(); // EMC - Plugins should have access to this
         }
 
         Plugin[] plugins = pluginManager.getPlugins();
@@ -324,7 +325,7 @@ public final class CraftServer implements Server {
             setVanillaCommands(false);
             // Spigot end
             commandMap.registerServerAliases();
-            loadCustomPermissions();
+            // loadCustomPermissions(); // EMC - Moving up
             DefaultPermissions.registerCorePermissions();
             CraftDefaultPermissions.registerCorePermissions();
             helpMap.initializeCommands();
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 211c421..46f10f6 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -125,6 +125,7 @@ public class CraftEventFactory {
         BlockPlaceEvent event = new BlockPlaceEvent(placedBlock, replacedBlockState, blockClicked, player.getItemInHand(), player, canBuild);
         craftServer.getPluginManager().callEvent(event);
 
+        if (event.isCancelled()) { player.updateInventory(); } // EMC
         return event;
     }
 
@@ -226,6 +227,7 @@ public class CraftEventFactory {
         }
         craftServer.getPluginManager().callEvent(event);
 
+        if (action == Action.RIGHT_CLICK_BLOCK && event.isCancelled()) { player.updateInventory(); } // EMC
         return event;
     }
 
-- 
1.9.1

